src/Sharpliner/AzureDevOps/PublicDefinitions.cs

```diff
@@ -71,6 +80,161 @@ public abstract class JobTemplateDefinition : TemplateDefinition<JobBase>
     public sealed override IReadOnlyCollection<IDefinitionValidation> Validations => Definition.GetJobValidations();
 }
 
+public interface ITemplateParametersProvider
+{
+    List<Parameter> ToParameters();
+
+    TemplateParameters ToTemplateParameters();
+}
+
+public abstract class TemplateParametersProviderBase<TParameters> : ITemplateParametersProvider where TParameters : class, new()
+{
+    public List<Parameter> ToParameters()
+    {
+        var result = new List<Parameter>();
+        var defaultParameters = new TParameters();
+        foreach (var property in typeof(TParameters).GetProperties())
+        {
+            var name = property.Name;
+            var yamlMember = property.GetCustomAttribute<YamlMemberAttribute>();
+            if (yamlMember?.Alias is not null)
+            {
+                name = yamlMember.Alias;
+            }
+
+            var defaultValue = property.GetValue(defaultParameters);
+            Parameter parameter;
+            if (property.PropertyType.IsEnum)
```

Comment:
hmm, I haven't tried yet, although I'm not sure how much cleaner it will be as we cannot use a constant expression here so we'll have to add a `when` condition to each case

Reply:
I think some pattern matching might do the trick:

```csharp
            Parameter parameter = property.PropertyType switch
            {
                { } type when type.IsEnum => new StringParameter(name, defaultValue: defaultValue as string, allowedValues: Enum.GetNames(property.PropertyType)),
                { } type when type == typeof(string) => new StringParameter(name, defaultValue: defaultValue as string),
                { } type when type == typeof(bool) => new BooleanParameter(name, defaultValue: defaultValue as bool?),
                { } type when type == typeof(int) => new NumberParameter(name, defaultValue: defaultValue as int?),
                { } type when type == typeof(Step) => new StepParameter(name, defaultValue: defaultValue as Step),
                { } type when type == typeof(ConditionedList<Step>) => new StepListParameter(name, defaultValue: defaultValue as ConditionedList<Step>),
                { } type when type.IsAssignableFrom(typeof(JobBase)) => new JobParameter(name, defaultValue: defaultValue as JobBase),
                { } type when type == typeof(ConditionedList<JobBase>) => new JobListParameter(name, defaultValue: defaultValue as ConditionedList<JobBase>),
                { } type when type == typeof(DeploymentJob) => new DeploymentParameter(name, defaultValue: defaultValue as DeploymentJob),
                { } type when type == typeof(ConditionedList<DeploymentJob>) => new DeploymentListParameter(name, defaultValue: defaultValue as ConditionedList<DeploymentJob>),
                { } type when type == typeof(Stage) => new StageParameter(name, defaultValue: defaultValue as Stage),
                { } type when type == typeof(ConditionedList<Stage>) => new StageListParameter(name, defaultValue: defaultValue as ConditionedList<Stage>),
                _ => new ObjectParameter(name),
            };
```